org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/CompilationUnitDocumentProvider.java

```diff
@@ -1617,6 +1617,9 @@ public void run() {
 							if (buffer.hasUnsavedChanges())
 								buffer.save(getSubProgressMonitor(monitor, 1), true);
 
+							if (stamp != unit.getResource().getModificationStamp()) {
+								unit.updateTimeStamp();
```

Comment:
Both new tests fail if this line is commented out, this shows the original design issue with `CompilationUnitDocumentProvider.commitWorkingCopy(IProgressMonitor, Object, CompilationUnitInfo, boolean)` .

The `CommitWorkingCopyOperation.executeOperation()` was the only code "touching" `CompilationUnitElementInfo.timestamp` before and it **was** called from `CompilationUnitDocumentProvider.commitWorkingCopy()` right **before** `CompilationUnitDocumentProvider.notifyPostSaveListeners()` started to call into "listeners" that were not just "listeners" but actively changing the content of the compilation unit... 

So the patch here is a workaround for **one** broken aspect. There could be more.